File: Easy/20_ValidParentheses.java
class Solution {
    public boolean isValid(String s) {
        // 먼저 짝꿍 맞추기
        Map<Character, Character> map = new HashMap<>();
        map.put(')', '(');
        map.put('}', '{');
        map.put(']', '[');
        
        // 스택 생성
        Stack<Character> stack = new Stack();
        for (Character ch : s.toCharArray()) {
            if (!map.keySet().contains(ch)) {
                stack.push(ch);
            } else {
                if (!stack.isEmpty()) {
                    if (stack.peek() == map.get(ch)) {
                        stack.pop();
                    } else {
                        return false;
                    }
                } else {
                    return false;
                }
            }
        }
        
        return stack.isEmpty();
    }
}

File: Easy/20_ValidParentheses.ts
function isValid(s: string): boolean {
    
    //     const map = new Map();
        
    //     map.set(')', '(');
    //     map.set('}', '{');
    //     map.set(']', '[');
        
        const map: {[key: string]: string} = {
            ')': '(',
            '}': '{',
            ']': '['
        }
        
        const stack: Array<string> = [];
        
        for (let i = 0; i < s.length; i++) {
            if (!Object.keys(map).includes(s.charAt(i))) {
                stack.push(s.charAt(i));
            } else {
                if (stack.length !== 0 && stack[stack.length - 1] === map[s.charAt(i)]) {
                    stack.pop();
                } else {
                    return false;
                }
            }
        }
        
        return stack.length === 0;
    };